metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,300 @@
+    if (require_accversion_to_taxid_map) {
+        logger->trace("Parsing label_taxid_map file...");
+        read_accversion_to_taxid_map(label_taxid_map_filepath, anno_matrix_);
+        logger->trace("Finished label_taxid_map file in {} sec", timer.elapsed());
+    }
+
+    timer.reset();
+    logger->trace("Parsing taxonomic tree...");
+    ChildrenList tree;
+    read_tree(tax_tree_filepath, &tree);
+    logger->trace("Finished taxonomic tree read in {} sec.", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Calculating tree statistics...");
+    std::vector<TaxId> tree_linearization;
+    dfs_statistics(root_node_, tree, &tree_linearization);
+    logger->trace("Finished tree statistics calculation in {} sec.", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Starting rmq preprocessing...");
+    rmq_preprocessing(tree_linearization);
+    logger->trace("Finished rmq preprocessing in {} sec.", timer.elapsed());
+}
+
+void TaxonomyClsAnno::read_tree(const std::string &tax_tree_filepath, ChildrenList *tree) {
+    std::ifstream f(tax_tree_filepath);
+    if (!f.good()) {
+        logger->error("Error: Failed to open Taxonomic Tree file {}", tax_tree_filepath);
+        exit(1);
+    }
+
+    std::string line;
+    tsl::hopscotch_map<TaxId, TaxId> full_parents_list;
+    while (getline(f, line)) {
+        if (line == "") {
+            logger->error("Error: The Taxonomic Tree file contains empty lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        std::vector<std::string> parts = utils::split_string(line, "\t");
+        if (parts.size() <= 2) {
+            logger->error("Error: The Taxonomic tree filepath contains incomplete lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        uint32_t act = std::stoul(parts[0]);
+        uint32_t parent = std::stoul(parts[2]);
+        full_parents_list[act] = parent;
+        node_parent_[act] = parent;
+    }
+
+    std::vector<TaxId> relevant_taxids;
+    // 'considered_relevant_taxids' is used to make sure that there are no duplications in 'relevant_taxids'.
+    tsl::hopscotch_set<TaxId> considered_relevant_taxids;
+
+    if (accversion_to_taxid_map_.size()) {
+        // Store only the taxonomic nodes that exists in the annotation matrix.
+        for (const std::pair<std::string, TaxId> &pair : accversion_to_taxid_map_) {
+            relevant_taxids.push_back(pair.second);
+            considered_relevant_taxids.insert(pair.second);
+        }
+    } else {
+        // If 'this->accversion_to_taxid_map' is empty, store the entire taxonomic tree.
+        for (auto it : full_parents_list) {
+            relevant_taxids.push_back(it.first);
+            considered_relevant_taxids.insert(it.first);
```

Comment:
Rename `it` -> `pair`.

`it` would be an iterator in the container.

Or, better write the variables explicitly:
`for (const auto &[???, ???] : full_parents_list) {`

Reply:
Also, I'm not sure `full_parents_list` is a good name... What does this thing store?

Reply:
I have merged together the `full_parents_list` and `node_parent_` to delete the entire `full_parents_list` object. In this way, I have also reduced the number of `hopscotch_map` accesses. Thanks for the suggestion!